color/src/css.rs

```diff
@@ -0,0 +1,222 @@
+            components: add_alpha(components, alpha),
+        }
+    }
+
+    fn premultiply_split(self) -> ([f32; 3], f32) {
+        // Reference: §12.3 of Color 4 spec
+        let (opaque, alpha) = split_alpha(self.components);
+        let premul = if alpha == 1.0 || self.missing.contains(3) {
+            opaque
+        } else {
+            self.cs.layout().scale(opaque, alpha)
+        };
+        (premul, alpha)
+    }
+
+    fn powerless_to_missing(&mut self) {
+        // Note: the spec seems vague on the details of what this should do,
+        // and there is some controversy in discussion threads. For example,
+        // in Lab-like spaces, if L is 0 do the other components become powerless?
+        const POWERLESS_EPSILON: f32 = 1e-6;
+        if self.cs.layout() != ColorspaceLayout::Rectangular
+            && self.components[1] < POWERLESS_EPSILON
+        {
+            self.cs
+                .set_h_missing(&mut self.missing, &mut self.components);
+        }
+    }
+
+    /// Interpolate two colors, according to CSS Color 4 spec.
+    ///
+    /// This method does a bunch of precomputation, resulting in an [`Interpolator`]
+    /// object that can be evaluated at various `t` values.
+    ///
+    /// Reference: §12 of Color 4 spec
+    pub fn interpolate(
+        self,
+        other: Self,
+        cs: ColorspaceTag,
+        direction: HueDirection,
+    ) -> Interpolator {
+        let mut a = self.convert(cs);
+        let mut b = other.convert(cs);
+        let missing = a.missing & b.missing;
+        if self.missing != other.missing {
+            for i in 0..4 {
+                if (a.missing & !b.missing).contains(i) {
+                    a.components[i] = b.components[i];
+                } else if (!a.missing & b.missing).contains(i) {
+                    b.components[i] = a.components[i];
+                }
+            }
+        }
+        let (premul1, alpha1) = a.premultiply_split();
+        let (mut premul2, alpha2) = b.premultiply_split();
+        fixup_hues_for_interpolate(premul1, &mut premul2, cs.layout(), direction);
+        Interpolator {
+            premul1,
+            alpha1,
+            premul2,
+            alpha2,
+            cs,
+            missing,
+        }
+    }
+
+    /// Compute the relative luminance of the color.
+    ///
+    /// This can be useful for choosing contrasting colors, and follows the
+    /// WCAG 2.1 spec.
+    ///
+    /// Note that this method only considers the opaque color, not the alpha.
+    /// Blending semi-transparent colors will reduce contrast, and that
+    /// should also be taken into account.
+    pub fn relative_luminance(self) -> f32 {
+        let rgb = self.convert(ColorspaceTag::LinearSrgb).components;
+        0.2126 * rgb[0] + 0.7152 * rgb[1] + 0.0722 * rgb[2]
+    }
+}
+
+impl Interpolator {
+    pub fn eval(&self, t: f32) -> CssColor {
+        let premul = [
+            self.premul1[0] + t * (self.premul2[0] - self.premul1[0]),
```

Comment:
It doesn't look like `premul2` is ever used except to compute the same difference value here.

Should this difference instead be stored?

Reply:
As a bit of super-geekery, there are two basic strategies for implementing lerp. One is `a + t * (b - a)`, the other is `a * (1 - t) + b * t`. The latter has the advantage of zero rounding error at t = 1, but I think for the purposes of this crate one ulp at f32 is completely acceptable.

Reply:
`glam` ended up switching to the precise lerp (reported in https://github.com/bitshifter/glam-rs/issues/275, fixed in https://github.com/bitshifter/glam-rs/pull/548 (but some of the concerns about differences in magnitude are different in color code vs generic vectors ...)

Reply:
Right, for geometry sometimes you really care about watertightness, but for color close is pretty much always good enough.